// @stoplight/spectral-rulesets/dist/asyncapi/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const spectral_formats_1 = require("@stoplight/spectral-formats");
const spectral_functions_1 = require("@stoplight/spectral-functions");
const asyncApi2SchemaValidation_1 = (0, tslib_1.__importDefault)(require("./functions/asyncApi2SchemaValidation"));
const asyncApi2PayloadValidation_1 = (0, tslib_1.__importDefault)(require("./functions/asyncApi2PayloadValidation"));
const asyncApi2Schema = (0, tslib_1.__importStar)(require("./schemas/schema.asyncapi2.json"));
exports.default = {
    documentationUrl: 'https://meta.stoplight.io/docs/spectral/docs/reference/asyncapi-rules.md',
    formats: [spectral_formats_1.asyncApi2],
    rules: {
        'asyncapi-channel-no-empty-parameter': {
            description: 'Channel path must not have empty parameter substitution pattern.',
            recommended: true,
            type: 'style',
            given: '$.channels',
            then: {
                field: '@key',
                function: spectral_functions_1.pattern,
                functionOptions: {
                    notMatch: '{}',
                },
            },
        },
        'asyncapi-channel-no-query-nor-fragment': {
            description: 'Channel path must not include query ("?") or fragment ("#") delimiter.',
            recommended: true,
            type: 'style',
            given: '$.channels',
            then: {
                field: '@key',
                function: spectral_functions_1.pattern,
                functionOptions: {
                    notMatch: '[\\?#]',
                },
            },
        },
        'asyncapi-channel-no-trailing-slash': {
            description: 'Channel path must not end with slash.',
            recommended: true,
            type: 'style',
            given: '$.channels',
            then: {
                field: '@key',
                function: spectral_functions_1.pattern,
                functionOptions: {
                    notMatch: '.+\\/$',
                },
            },
        },
        'asyncapi-headers-schema-type-object': {
            description: 'Headers schema type must be "object".',
            message: 'Headers schema type must be "object" ({{error}}).',
            severity: 'error',
            recommended: true,
            type: 'validation',
            given: [
                '$.components.messageTraits.*.headers',
                '$.components.messages.*.headers',
                '$.channels.*.[publish,subscribe].message.headers',
            ],
            then: {
                function: spectral_functions_1.schema,
                functionOptions: {
                    allErrors: true,
                    schema: {
                        type: 'object',
                        properties: {
                            type: {
                                enum: ['object'],
                            },
                        },
                        required: ['type'],
                    },
                },
            },
        },
        'asyncapi-info-contact-properties': {
            description: 'Contact object must have "name", "url" and "email".',
            recommended: true,
            type: 'style',
            given: '$.info.contact',
            then: [
                {
                    field: 'name',
                    function: spectral_functions_1.truthy,
                },
                {
                    field: 'url',
                    function: spectral_functions_1.truthy,
                },
                {
                    field: 'email',
                    function: spectral_functions_1.truthy,
                },
            ],
        },
        'asyncapi-info-contact': {
            description: 'Info object must have "contact" object.',
            recommended: true,
            type: 'style',
            given: '$',
            then: {
                field: 'info.contact',
                function: spectral_functions_1.truthy,
            },
        },
        'asyncapi-info-description': {
            description: 'Info "description" must be present and non-empty string.',
            recommended: true,
            type: 'style',
            given: '$',
            then: {
                field: 'info.description',
                function: spectral_functions_1.truthy,
            },
        },
        'asyncapi-info-license-url': {
            description: 'License object must include "url".',
            recommended: false,
            type: 'style',
            given: '$',
            then: {
                field: 'info.license.url',
                function: spectral_functions_1.truthy,
            },
        },
        'asyncapi-info-license': {
            description: 'Info object must have "license" object.',
            recommended: true,
            type: 'style',
            given: '$',
            then: {
                field: 'info.license',
                function: spectral_functions_1.truthy,
            },
        },
        'asyncapi-operation-description': {
            description: 'Operation "description" must be present and non-empty string.',
            recommended: true,
            type: 'style',
            given: '$.channels[*][publish,subscribe]',
            then: {
                field: 'description',
                function: spectral_functions_1.truthy,
            },
        },
        'asyncapi-operation-operationId': {
            description: 'Operation must have "operationId".',
            severity: 'error',
            recommended: true,
            type: 'validation',
            given: '$.channels[*][publish,subscribe]',
            then: {
                field: 'operationId',
                function: spectral_functions_1.truthy,
            },
        },
        'asyncapi-parameter-description': {
            description: 'Parameter objects must have "description".',
            recommended: false,
            type: 'style',
            given: ['$.components.parameters.*', '$.channels.*.parameters.*'],
            then: {
                field: 'description',
                function: spectral_functions_1.truthy,
            },
        },
        'asyncapi-payload-default': {
            description: 'Default must be valid against its defined schema.',
            message: '{{error}}',
            severity: 'error',
            recommended: true,
            type: 'validation',
            given: [
                '$.components.messageTraits[?(@.schemaFormat === void 0)].payload.default^',
                '$.components.messages[?(@.schemaFormat === void 0)].payload.default^',
                "$.channels[*][publish,subscribe][?(@property === 'message' && @.schemaFormat === void 0)].payload.default^",
            ],
            then: {
                function: asyncApi2SchemaValidation_1.default,
                functionOptions: {
                    type: 'default',
                },
            },
        },
        'asyncapi-payload-examples': {
            description: 'Examples must be valid against their defined schema.',
            message: '{{error}}',
            severity: 'error',
            recommended: true,
            type: 'validation',
            given: [
                '$.components.messageTraits[?(@.schemaFormat === void 0)].payload.examples^',
                '$.components.messages[?(@.schemaFormat === void 0)].payload.examples^',
                "$.channels[*][publish,subscribe][?(@property === 'message' && @.schemaFormat === void 0)].payload.examples^",
            ],
            then: {
                function: asyncApi2SchemaValidation_1.default,
                functionOptions: {
                    type: 'examples',
                },
            },
        },
        'asyncapi-payload-unsupported-schemaFormat': {
            description: 'Message schema validation is only supported with default unspecified "schemaFormat".',
            severity: 'info',
            recommended: true,
            type: 'validation',
            given: ['$.components.messageTraits.*', '$.components.messages.*', '$.channels[*][publish,subscribe].message'],
            then: {
                field: 'schemaFormat',
                function: spectral_functions_1.undefined,
            },
        },
        'asyncapi-payload': {
            description: 'Payloads must be valid against AsyncAPI Schema object.',
            message: '{{error}}',
            severity: 'error',
            recommended: true,
            type: 'validation',
            given: [
                '$.components.messageTraits[?(@.schemaFormat === void 0)].payload',
                '$.components.messages[?(@.schemaFormat === void 0)].payload',
                "$.channels[*][publish,subscribe][?(@property === 'message' && @.schemaFormat === void 0)].payload",
            ],
            then: {
                function: asyncApi2PayloadValidation_1.default,
            },
        },
        'asyncapi-schema-default': {
            description: 'Default must be valid against its defined schema.',
            message: '{{error}}',
            severity: 'error',
            recommended: true,
            type: 'validation',
            given: [
                '$.components.schemas.*.default^',
                '$.components.parameters.*.schema.default^',
                '$.channels.*.parameters.*.schema.default^',
            ],
            then: {
                function: asyncApi2SchemaValidation_1.default,
                functionOptions: {
                    type: 'default',
                },
            },
        },
        'asyncapi-schema-examples': {
            description: 'Examples must be valid against their defined schema.',
            message: '{{error}}',
            severity: 'error',
            recommended: true,
            type: 'validation',
            given: [
                '$.components.schemas.*.examples^',
                '$.components.parameters.*.schema.examples^',
                '$.channels.*.parameters.*.schema.examples^',
            ],
            then: {
                function: asyncApi2SchemaValidation_1.default,
                functionOptions: {
                    type: 'examples',
                },
            },
        },
        'asyncapi-schema': {
            description: 'Validate structure of AsyncAPI v2.0.0 Specification.',
            message: '{{error}}',
            severity: 'error',
            recommended: true,
            type: 'validation',
            given: '$',
            then: {
                function: spectral_functions_1.schema,
                functionOptions: {
                    allErrors: true,
                    schema: asyncApi2Schema,
                },
            },
        },
        'asyncapi-server-no-empty-variable': {
            description: 'Server URL must not have empty variable substitution pattern.',
            recommended: true,
            type: 'style',
            given: '$.servers[*].url',
            then: {
                function: spectral_functions_1.pattern,
                functionOptions: {
                    notMatch: '{}',
                },
            },
        },
        'asyncapi-server-no-trailing-slash': {
            description: 'Server URL must not end with slash.',
            recommended: true,
            type: 'style',
            given: '$.servers[*].url',
            then: {
                function: spectral_functions_1.pattern,
                functionOptions: {
                    notMatch: '/$',
                },
            },
        },
        'asyncapi-server-not-example-com': {
            description: 'Server URL must not point at example.com.',
            recommended: false,
            type: 'style',
            given: '$.servers[*].url',
            then: {
                function: spectral_functions_1.pattern,
                functionOptions: {
                    notMatch: 'example\\.com',
                },
            },
        },
        'asyncapi-servers': {
            description: 'AsyncAPI object must have non-empty "servers" object.',
            recommended: true,
            type: 'validation',
            given: '$',
            then: {
                field: 'servers',
                function: spectral_functions_1.schema,
                functionOptions: {
                    schema: {
                        type: 'object',
                        minProperties: 1,
                    },
                    allErrors: true,
                },
            },
        },
        'asyncapi-tag-description': {
            description: 'Tag object must have "description".',
            recommended: false,
            type: 'style',
            given: '$.tags[*]',
            then: {
                field: 'description',
                function: spectral_functions_1.truthy,
            },
        },
        'asyncapi-tags-alphabetical': {
            description: 'AsyncAPI object must have alphabetical "tags".',
            recommended: false,
            type: 'style',
            given: '$',
            then: {
                field: 'tags',
                function: spectral_functions_1.alphabetical,
                functionOptions: {
                    keyedBy: 'name',
                },
            },
        },
        'asyncapi-tags': {
            description: 'AsyncAPI object must have non-empty "tags" array.',
            recommended: true,
            type: 'style',
            given: '$',
            then: {
                field: 'tags',
                function: spectral_functions_1.truthy,
            },
        },
        'asyncapi-unused-components-schema': {
            description: 'Potentially unused components schema has been detected.',
            recommended: true,
            type: 'style',
            resolved: false,
            given: '$.components.schemas',
            then: {
                function: spectral_functions_1.unreferencedReusableObject,
                functionOptions: {
                    reusableObjectsLocation: '#/components/schemas',
                },
            },
        },
    },
};
