// @stoplight/spectral-rulesets/dist/oas/functions/oasOpSecurityDefined.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const spectral_core_1 = require("@stoplight/spectral-core");
const getAllOperations_1 = require("./utils/getAllOperations");
const isObject_1 = require("./utils/isObject");
function _get(value, path) {
    for (const segment of path) {
        if (!(0, isObject_1.isObject)(value)) {
            break;
        }
        value = value[segment];
    }
    return value;
}
exports.default = (0, spectral_core_1.createRulesetFunction)({
    input: {
        type: 'object',
        properties: {
            paths: {
                type: 'object',
            },
        },
    },
    options: {
        type: 'object',
        properties: {
            schemesPath: {
                type: 'array',
                items: {
                    type: ['string', 'number'],
                },
            },
        },
    },
}, function oasOpSecurityDefined(targetVal, { schemesPath }) {
    const { paths } = targetVal;
    const results = [];
    const schemes = _get(targetVal, schemesPath);
    const allDefs = (0, isObject_1.isObject)(schemes) ? Object.keys(schemes) : [];
    for (const { path, operation, value } of (0, getAllOperations_1.getAllOperations)(paths)) {
        if (!(0, isObject_1.isObject)(value))
            continue;
        const { security } = value;
        if (!Array.isArray(security)) {
            continue;
        }
        for (const [index, value] of security.entries()) {
            if (!(0, isObject_1.isObject)(value)) {
                continue;
            }
            const securityKeys = Object.keys(value);
            if (securityKeys.length > 0 && !allDefs.includes(securityKeys[0])) {
                results.push({
                    message: 'Operation must not reference an undefined security scheme.',
                    path: ['paths', path, operation, 'security', index],
                });
            }
        }
    }
    return results;
});
